Mathlib/Tactic/NormNum/BigOperators.lean

```diff
@@ -0,0 +1,372 @@
+    let h : Q(Multiset.Nodup ($a ::ₘ $s')) := q($pf ▸ $nd)
+    let nd' : Q(Multiset.Nodup $s') := q((Multiset.nodup_cons.mp $h).2)
+    let h' : Q($a ∉ $s') := q((Multiset.nodup_cons.mp $h).1)
+    return (.cons a q(Finset.mk $s' $nd') h'
+      (q($pf ▸ Finset.mk_cons $h) : Q(Finset.mk $val $nd = Finset.cons $a ⟨$s', $nd'⟩ $h')))
+| (`Finset.range, #[(n : Q(ℕ))]) => do
+  match ← Nat.unifyZeroOrSucc n with
+  | .zero _pf => do
+    pure (.empty (q(Finset.range_zero) : Q(Finset.range 0 = ∅)))
+  | .succ n' _pf => pure <| ((Finset.ProveEmptyOrConsResult.cons
+      n'
+      (q(Finset.range $n'))
+      (q(@Finset.not_mem_range_self $n'))
+      (q(Finset.range_succ_eq_cons $n'))).uncheckedCast
+        (q(Finset.range (Nat.succ $n')) : Q(Finset ℕ))
+        s)
+| (`Finset.univ, #[_, instFT]) => do
+  match Expr.getAppFnArgs (← whnfI instFT) with
+  | (`Fintype.mk, #[_, (elems : Q(Finset $α)), (complete : Q(∀ x : $α, x ∈ $elems))]) => do
+    have _instFT : Q(Fintype $α) := instFT
+    let res ← Finset.proveEmptyOrCons elems
+    pure <| res.eq_trans (q(Finset.univ_eq_elems $elems $complete) : Q(Finset.univ = $elems))
+  | e => throwError "Finset.proveEmptyOrCons: could not determine elements of Fintype instance {e}"
+| (fn, args) =>
+  throwError "Finset.proveEmptyOrCons: unsupported finset expression {s} ({fn}, {args})"
+
+namespace NormNum
+
+/-- If `a = b` and we can evaluate `b`, then we can evaluate `a`. -/
+def Result.eq_trans {α : Q(Type u)} {a b : Q($α)} (eq : Q($a = $b)) : Result b → Result a
+| .isBool true proof =>
+  have a : Q(Prop) := a
+  have b : Q(Prop) := b
+  have eq : Q($a = $b) := eq
+  have proof : Q($b) := proof
+  Result.isTrue (x := a) q($eq ▸ $proof)
+| .isBool false proof =>
+  have a : Q(Prop) := a
+  have b : Q(Prop) := b
+  have eq : Q($a = $b) := eq
+  have proof : Q(¬ $b) := proof
+ Result.isFalse (x := a) q($eq ▸ $proof)
+| .isNat inst lit proof => Result.isNat inst lit q($eq ▸ $proof)
+| .isNegNat inst lit proof => Result.isNegNat inst lit q($eq ▸ $proof)
+| .isRat inst q n d proof => Result.isRat inst q n d q($eq ▸ $proof)
+
+protected lemma Finset.sum_empty {β α : Type _} [CommSemiring β] (f : α → β) :
+    IsNat (Finset.sum ∅ f) 0 :=
+  ⟨by simp⟩
+
+protected lemma Finset.prod_empty {β α : Type _} [CommSemiring β] (f : α → β) :
+    IsNat (Finset.prod ∅ f) 1 :=
+  ⟨by simp⟩
+
+/-- Evaluate a big operator applied to a finset by repeating `proveEmptyOrCons` until
+we exhaust all elements of the set. -/
+partial def evalFinsetBigop {α : Q(Type u)} {β : Q(Type v)}
+    (op : Q(Finset $α → ($α → $β) → $β))
+    (f : Q($α → $β))
+    (res_empty : Result q($op Finset.empty $f))
+    (res_cons : {a : Q($α)} -> {s' : Q(Finset $α)} -> {h : Q($a ∉ $s')} ->
+      Result (α := β) q($f $a) -> Result (α := β) q($op $s' $f) ->
+      MetaM (Result (α := β) q($op (Finset.cons $a $s' $h) $f))) :
+    (s : Q(Finset $α)) → MetaM (Result (α := β) q($op $s $f))
+| s => do
+  match ← Finset.proveEmptyOrCons s with
```

Comment:
Just thinking out loud here, not necessarily a request for this implementation:

Could we have functions like `List.proveEq` (perhaps the wrong name) that takes a `Q(List α)` and returns a `List (Q α)` and an appropriate proof? This (and similarly for Multiset/Finset) feel like they might be good building blocks for other tactics.

Reply:
This was my initial design for this tactic, and I think it would still work, however I could not figure out how to keep track of the proof that the `List (Q α)` indeed corresponds to the `Q(List α)` after we e.g. remove the first cons from the list.

A small additional advantage to the `proveConsOrNil` pattern is that it is a bit more lazy, so it might help performance slightly.